src/persist-client/src/usage.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Introspection of storage utilization by persist
+
+use std::sync::Arc;
+
+use crate::BlobKeyPrefix;
+use mz_persist::location::{Blob, ExternalError};
+
+use crate::cache::PersistClientCache;
+
+/// Provides access to storage usage metrics for a specific Blob
+#[derive(Debug)]
+pub struct StorageUsageClient {
+    blob: Arc<dyn Blob + Send + Sync>,
+}
+
+impl StorageUsageClient {
+    /// Creates a new StorageUsageClient pointed to a specific Blob
+    pub async fn open(
+        blob_uri: String,
+        client_cache: &mut PersistClientCache,
+    ) -> Result<StorageUsageClient, ExternalError> {
+        let blob = client_cache.open_blob(blob_uri).await?;
+        Ok(StorageUsageClient { blob })
+    }
+
+    /// Returns the size (in bytes) of a subset of blobs specified by [crate::BlobKeyPrefix]
+    pub async fn size(&self, prefix: BlobKeyPrefix<'_>) -> Result<u64, ExternalError> {
+        let mut total_size = 0;
+        self.blob
```

Comment:
in general, persist tries very hard to internally retry errors whenever that wouldn't lead to a correctness issue, so let's put a retry_external loop here so we don't have to return an error. (seems like it would have to retry the entire method, otherwise total_size might double count things)

Reply:
ooo it is just dawning on me that `retry_external` converts a `Result<R, ExternalError>` into a `R`! that's really cool, love the use of error types to bubble this up